{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import requests\n",
    "import gzip\n",
    "import sys\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from statsmodels.stats.rates import test_poisson, confint_poisson"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# list of GPCR genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2056\n",
      "423\n",
      "397\n",
      "397\n"
     ]
    }
   ],
   "source": [
    "receptors = requests.get('https://gpcrdb.org/services/receptorlist/').json()\n",
    "print(len(receptors))\n",
    "human_receptors = [receptor for receptor in receptors \n",
    "    if (receptor['species'] == 'Homo sapiens')]\n",
    "print(len(human_receptors))\n",
    "human_nonolf_receptors = [receptor for receptor in human_receptors \n",
    "    if (receptor['receptor_family'] not in ['Olfactory'])]\n",
    "print(len(human_nonolf_receptors))\n",
    "human_receptors = pd.DataFrame(human_nonolf_receptors)\n",
    "human_receptors = human_receptors[~human_receptors.entry_name.duplicated()]\n",
    "print(human_receptors.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_receptors_ref = pd.read_csv('../data/labels/gene_families/gpcr_genes_human_gpcrdb.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry_name</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>receptor_class</th>\n",
       "      <th>receptor_family</th>\n",
       "      <th>ligand_type</th>\n",
       "      <th>subfamily</th>\n",
       "      <th>endogenous_ligands</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>npy6r_human</td>\n",
       "      <td>y&lt;sub&gt;6&lt;/sub&gt; receptor</td>\n",
       "      <td>Q99463</td>\n",
       "      <td>Class A (Rhodopsin)</td>\n",
       "      <td>Neuropeptide Y receptors</td>\n",
       "      <td>Peptide receptors</td>\n",
       "      <td>y&lt;sub&gt;6&lt;/sub&gt; receptor</td>\n",
       "      <td>[]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>gpr33_human</td>\n",
       "      <td>&lt;i&gt;GPR33&lt;/i&gt;</td>\n",
       "      <td>Q49SQ1</td>\n",
       "      <td>Class A (Rhodopsin)</td>\n",
       "      <td>Class A orphans</td>\n",
       "      <td>Orphan receptors</td>\n",
       "      <td>&lt;i&gt;GPR33&lt;/i&gt;</td>\n",
       "      <td>[]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>taar3_human</td>\n",
       "      <td>&lt;i&gt;TAAR3&lt;/i&gt;</td>\n",
       "      <td>Q9P1P4</td>\n",
       "      <td>Class A (Rhodopsin)</td>\n",
       "      <td>Class A orphans</td>\n",
       "      <td>Orphan receptors</td>\n",
       "      <td>&lt;i&gt;TAAR3&lt;/i&gt;</td>\n",
       "      <td>[{'name': 'isoamylamine'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>taar9_human</td>\n",
       "      <td>&lt;i&gt;TAAR9&lt;/i&gt;</td>\n",
       "      <td>Q96RI9</td>\n",
       "      <td>Class A (Rhodopsin)</td>\n",
       "      <td>Class A orphans</td>\n",
       "      <td>Orphan receptors</td>\n",
       "      <td>&lt;i&gt;TAAR9&lt;/i&gt;</td>\n",
       "      <td>[]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>agre4_human</td>\n",
       "      <td>ADGRE4P</td>\n",
       "      <td>Q86SQ3</td>\n",
       "      <td>Class B2 (Adhesion)</td>\n",
       "      <td>ADGRE</td>\n",
       "      <td>Adhesion receptors</td>\n",
       "      <td>ADGRE4P</td>\n",
       "      <td>[]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>t2r45_human</td>\n",
       "      <td>&lt;i&gt;TAS2R45&lt;/i&gt;</td>\n",
       "      <td>P59539</td>\n",
       "      <td>Class T (Taste 2)</td>\n",
       "      <td>Taste 2 receptors</td>\n",
       "      <td>Sensory receptors</td>\n",
       "      <td>&lt;i&gt;TAS2R45&lt;/i&gt;</td>\n",
       "      <td>[]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      entry_name                    name accession       receptor_class  \\\n",
       "79   npy6r_human  y<sub>6</sub> receptor    Q99463  Class A (Rhodopsin)   \n",
       "222  gpr33_human            <i>GPR33</i>    Q49SQ1  Class A (Rhodopsin)   \n",
       "282  taar3_human            <i>TAAR3</i>    Q9P1P4  Class A (Rhodopsin)   \n",
       "286  taar9_human            <i>TAAR9</i>    Q96RI9  Class A (Rhodopsin)   \n",
       "310  agre4_human                 ADGRE4P    Q86SQ3  Class B2 (Adhesion)   \n",
       "388  t2r45_human          <i>TAS2R45</i>    P59539    Class T (Taste 2)   \n",
       "\n",
       "              receptor_family         ligand_type               subfamily  \\\n",
       "79   Neuropeptide Y receptors   Peptide receptors  y<sub>6</sub> receptor   \n",
       "222           Class A orphans    Orphan receptors            <i>GPR33</i>   \n",
       "282           Class A orphans    Orphan receptors            <i>TAAR3</i>   \n",
       "286           Class A orphans    Orphan receptors            <i>TAAR9</i>   \n",
       "310                     ADGRE  Adhesion receptors                 ADGRE4P   \n",
       "388         Taste 2 receptors   Sensory receptors          <i>TAS2R45</i>   \n",
       "\n",
       "             endogenous_ligands       species  \n",
       "79                           []  Homo sapiens  \n",
       "222                          []  Homo sapiens  \n",
       "282  [{'name': 'isoamylamine'}]  Homo sapiens  \n",
       "286                          []  Homo sapiens  \n",
       "310                          []  Homo sapiens  \n",
       "388                          []  Homo sapiens  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_receptors[human_receptors.entry_name.isin(pseudogenes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudogenes = [name for name in human_receptors.entry_name if name not in human_receptors_ref.entry_name.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      5ht1a_human\n",
       "1      5ht1b_human\n",
       "2      5ht1d_human\n",
       "3      5ht1e_human\n",
       "4      5ht1f_human\n",
       "          ...     \n",
       "392    gp107_human\n",
       "393    g137a_human\n",
       "394    tpra1_human\n",
       "395    gp143_human\n",
       "396    gp157_human\n",
       "Name: entry_name, Length: 397, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all human receptors from GPCRdb\n",
    "receptors = requests.get('https://gpcrdb.org/services/receptorlist/').json()\n",
    "human_receptors = [receptor for receptor in receptors \\\n",
    "    if (receptor['species'] == 'Homo sapiens') & \\\n",
    "        (receptor['receptor_family'] not in ['Olfactory receptors'])\n",
    "        ]\n",
    "human_receptors = pd.DataFrame(human_receptors)\n",
    "human_receptors['is_orphan'] = human_receptors.endogenous_ligands.map(len) == 0\n",
    "human_receptors = human_receptors[\n",
    "    ['entry_name',\n",
    "    'name',\n",
    "    'accession',\n",
    "    'receptor_class',\n",
    "    'receptor_family',\n",
    "    'ligand_type',\n",
    "    'subfamily',\n",
    "    'is_orphan'\n",
    "    ]]\n",
    "\n",
    "# Annotate with genes and coordinates using Uniprot Proteins API\n",
    "def get_genes(uniprot_accession):\n",
    "    requestURL = f\"https://www.ebi.ac.uk/proteins/api/proteins/{uniprot_accession}\"\n",
    "\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    responseBody = r.json()\n",
    "    return responseBody['gene']\n",
    "\n",
    "\n",
    "genes = []\n",
    "for uniprot_accession in tqdm(human_receptors.accession):\n",
    "    genes.append(get_genes(uniprot_accession))\n",
    "\n",
    "# Annotate with gene names and synonyms\n",
    "human_receptors['gene'] = [gene[0]['name']['value'] for gene in genes]\n",
    "\n",
    "gene_names = []\n",
    "for gene in genes:\n",
    "    primary_name = gene[0]['name']['value']\n",
    "    if 'synonyms' in gene[0].keys():\n",
    "        synonyms = [synonym['value'] for synonym in gene[0]['synonyms']]\n",
    "        gene_names.append([primary_name] + synonyms)\n",
    "    else:\n",
    "        gene_names.append(primary_name)\n",
    "human_receptors['gene_names'] = gene_names\n",
    "\n",
    "human_receptors = human_receptors[~human_receptors.entry_name.duplicated()]\n",
    "\n",
    "human_receptors.to_csv('../data/gpcr_genes_human_gpcrdb.tsv','\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constraint from gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad_constraint_path = \"../data/all_genes_constraint_gnomad.tsv\"\n",
    "r = requests.get(\"https://storage.googleapis.com/gcp-public-data--gnomad/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz\",stream=True)\n",
    "if r.status_code == 200:\n",
    "    with open(gnomad_constraint_path,'wb') as f:\n",
    "        r.raw.decode_content = True  # just in case transport encoding was applied\n",
    "        gzip_file = gzip.GzipFile(fileobj=r.raw)\n",
    "        shutil.copyfileobj(gzip_file, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get constraint for all gene regions in gnomad from gnomad summary statistics\n",
    "gnomad_constraint_path = \"../data/all_genes_constraint_gnomad.tsv\"\n",
    "if not os.path.exists(gnomad_constraint_path):\n",
    "    r = requests.get(\"https://storage.googleapis.com/gcp-public-data--gnomad/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz\",stream=True)\n",
    "    if r.status_code == 200:\n",
    "        with open(gnomad_constraint_path,'wb') as f:\n",
    "            r.raw.decode_content = True  # just in case transport encoding was applied\n",
    "            gzip_file = gzip.GzipFile(fileobj=r.raw)\n",
    "            shutil.copyfileobj(gzip_file, f)\n",
    "\n",
    "\n",
    "def clean_names(names):\n",
    "    try:\n",
    "        return eval(names)\n",
    "    except:\n",
    "        return names\n",
    "\n",
    "# Match to genes in gnomAD constraint calculations - check that none are missed\n",
    "gpcr_human_genes = pd.read_csv('../data/gpcr_genes_human_gpcrdb.tsv',sep='\\t',converters={'gene_names': clean_names})\n",
    "gnomad_constraint = pd.read_csv(gnomad_constraint_path,sep='\\t')\n",
    "\n",
    "gene_name_lookup = gpcr_human_genes[['gene','gene_names']].explode(column='gene_names')\n",
    "problematic_mappings = pd.DataFrame([\n",
    "    ('ADRA1D', 'ADRA1A'),\n",
    "    ('CCRL2', 'CCR6'),\n",
    "    ('ACKR2','CCR10'),\n",
    "    ('S1PR3','C9orf47'),\n",
    "    ('PTGER4' ,'PTGER2'),\n",
    "    ('FZD9' ,'FZD3')],columns =['gene','gene_names'])\n",
    "gene_name_lookup = problematic_mappings.merge(gene_name_lookup,how='outer',indicator=True)\n",
    "gene_name_lookup = gene_name_lookup[gene_name_lookup._merge != 'both'].drop(columns=['_merge'])\n",
    "gnomad_constraint = gene_name_lookup.merge(gnomad_constraint,left_on='gene_names',right_on='gene',suffixes=['','_gnomad'])\n",
    "gnomad_constraint = gnomad_constraint[gnomad_constraint.transcript != 'ENST00000528054'] # remove non-canonical transcripts\n",
    "\n",
    "# Select columns needed for analysis\n",
    "gnomad_constraint = gnomad_constraint[[\n",
    "    'gene','gene_gnomad','transcript',\n",
    "    'obs_lof','exp_lof',\n",
    "    'obs_mis','exp_mis',\n",
    "    'obs_mis_pphen','exp_mis_pphen',\n",
    "    'obs_syn','exp_syn',\n",
    "    'pLI'\n",
    "]]\n",
    "gnomad_constraint['obs_mis_non_pphen'] = gnomad_constraint.obs_mis - gnomad_constraint.obs_mis_pphen\n",
    "gnomad_constraint['exp_mis_non_pphen'] = gnomad_constraint.exp_mis - gnomad_constraint.exp_mis_pphen\n",
    "\n",
    "def poisson_test(obs, exp):\n",
    "    test = test_poisson(\n",
    "        count = obs, \n",
    "        nobs=1,\n",
    "        value=exp,\n",
    "        method=\"exact-c\",\n",
    "        alternative = \"smaller\"\n",
    "    )\n",
    "    ci = confint_poisson(\n",
    "        count = obs, \n",
    "        exposure=exp,\n",
    "        method=\"exact-c\"\n",
    "    )\n",
    "    return [obs/exp, ci[0], ci[1], np.log(test.pvalue)]\n",
    "\n",
    "for impact in ['syn','mis_non_pphen','mis_pphen','lof']:\n",
    "    gnomad_constraint[[f'oe_{impact}',f'oelf_{impact}',f'oeuf_{impact}',f'logp_{impact}']] = (\n",
    "        pd.DataFrame(\n",
    "            gnomad_constraint\n",
    "                .apply(lambda x: poisson_test(x[f'obs_{impact}'], x[f'exp_{impact}']),axis=1)\n",
    "            .tolist()\n",
    "            ))\n",
    "\n",
    "gnomad_constraint = gnomad_constraint[[\n",
    "    'gene','gene_gnomad','transcript', \n",
    "    'obs_lof','exp_lof','oe_lof','oelf_lof','oeuf_lof', 'logp_lof',\n",
    "    'obs_mis_pphen','exp_mis_pphen','oe_mis_pphen','oelf_mis_pphen','oeuf_mis_pphen', 'logp_mis_pphen',\n",
    "    'obs_mis_non_pphen','exp_mis_non_pphen','oe_mis_non_pphen','oelf_mis_non_pphen','oeuf_mis_non_pphen', 'logp_mis_non_pphen',\n",
    "    'obs_syn','exp_syn','oe_syn','oelf_syn','oeuf_syn', 'logp_syn',\n",
    "    'pLI'\n",
    "]]\n",
    "gnomad_constraint.to_csv('../data/gpcr_genes_constraint_gnomad.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add constraint from custom analysis\n",
    "custom_constraint = pd.read_csv('../data/gpcr_genes_custom_constraint.tsv')\n",
    "print(human_receptors.gene_gnomad.isna().sum())\n",
    "\n",
    "human_receptors.merge(custom_constraint, on = 'gene_gnomad')\n",
    "human_receptors.to_csv('../data/data/gpcr_genes_constraint_custom.txt')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Essential genes from MacArthur lab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_names(names):\n",
    "    try:\n",
    "        return eval(names)\n",
    "    except:\n",
    "        return names\n",
    "\n",
    "gpcr_human_genes = pd.read_csv('../data/gpcr_genes_human_gpcrdb.tsv',sep='\\t',converters={'gene_names': clean_names})\n",
    "#gnomad_constraint = pd.read_csv(gnomad_constraint_path,sep='\\t')\n",
    "\n",
    "gene_name_lookup = gpcr_human_genes[['gene','gene_names']].explode(column='gene_names')\n",
    "problematic_mappings = pd.DataFrame([\n",
    "    ('ADRA1D', 'ADRA1A'),\n",
    "    ('CCRL2', 'CCR6'),\n",
    "    ('ACKR2','CCR10'),\n",
    "    ('S1PR3','C9orf47'),\n",
    "    ('PTGER4' ,'PTGER2'),\n",
    "    ('FZD9' ,'FZD3')],columns =['gene','gene_names'])\n",
    "\n",
    "\n",
    "with open('../data/gene_lists/lists/mgi_essential.tsv','r') as fid:\n",
    "    mgi_essential_genes = [x.rstrip() for x in fid.readlines()]\n",
    "\n",
    "gene_name_lookup['mouse_essential'] = gene_name_lookup.gene_names.isin(mgi_essential_genes)\n",
    "\n",
    "with open('../data/gene_lists/lists/clinvar_path_likelypath.tsv','r') as fid:\n",
    "    clinvar_disease_genes = [x.rstrip() for x in fid.readlines()]\n",
    "gene_name_lookup['disease'] = gene_name_lookup.gene_names.isin(clinvar_disease_genes)\n",
    "\n",
    "with open('../data/gene_lists/lists/gwascatalog.tsv','r') as fid:\n",
    "    gwas_catalog_genes = [x.rstrip() for x in fid.readlines()]\n",
    "gene_name_lookup['gwas'] = gene_name_lookup.gene_names.isin(gwas_catalog_genes)\n",
    "\n",
    "genes_with_phenotypes = (\n",
    "    gene_name_lookup.groupby('gene')\n",
    "    .aggregate(np.any)\n",
    "    .drop(columns = ['gene_names'])\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "curated_mgi_phenotypes = pd.read_csv('../data/gpcr_genes_mgi_essential_curated.tsv',sep='\\t',index_col=0)\n",
    "curated_mgi_phenotypes.columns = ['gene','curated_mouse_phenotype']\n",
    "genes_with_phenotypes = genes_with_phenotypes.merge(curated_mgi_phenotypes,on='gene')\n",
    "\n",
    "genes_with_phenotypes.to_csv('../data/gpcrs_with_phenotypes.tsv',sep='\\t')\n",
    "# print(gene_name_lookup.essential.sum())\n",
    "# print(gene_name_lookup.disease.sum())\n",
    "# print(gene_name_lookup.gwas.sum())\n",
    "\n",
    "# print((gene_name_lookup.essential & gene_name_lookup.disease).sum())\n",
    "# print((gene_name_lookup.essential | gene_name_lookup.disease).sum())\n",
    "# print((gene_name_lookup.essential | gene_name_lookup.disease | gene_name_lookup.gwas).sum())\n",
    "\n",
    "#gene_name_lookup\n",
    "\n",
    "\n",
    "# Differences with curated version???"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curated drug targets from GlobalData\n",
    "\n",
    "* Check these for updates (FDA approvals) - none since 06/2022\n",
    "* Look for adverse event indications of drugs using DrugBank?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "gpcr_human_genes = pd.read_csv('../data/gpcr_genes_human_gpcrdb.tsv',sep='\\t',converters={'gene_names': clean_names})\n",
    "drug_targets = pd.read_csv('../data/gpcr_drug_targets.tsv',sep='\\t')\n",
    "gpcr_drug_targets = gpcr_human_genes.merge(drug_targets, left_on = 'entry_name',right_on ='target_uniprot',how='left')\n",
    "gpcr_drug_targets = gpcr_drug_targets[['gene','entry_name','approved_drugs_moa','adverse_event_associations']]\n",
    "gpcr_drug_targets['approved_drugs_moa'] = gpcr_drug_targets.approved_drugs_moa.fillna('none')\n",
    "gpcr_drug_targets['adverse_event_associations'] = gpcr_drug_targets.adverse_event_associations.fillna('No')\n",
    "gpcr_drug_targets.to_csv('../data/gpcrs_with_approved_drug_target_status.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make supplementary information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "gpcrs_with_families = pd.read_csv('../data/gpcr_genes_human_gpcrdb.tsv',sep='\\t')\n",
    "gpcrs_with_constraint = pd.read_csv('../data/gpcr_genes_constraint_gnomad.tsv',sep='\\t',index_col=0)\n",
    "gpcrs_with_phenotypes = pd.read_csv('../data/gpcrs_with_phenotypes.tsv',sep='\\t',index_col=0)\n",
    "gpcrs_with_drug_targets = pd.read_csv('../data/gpcrs_with_approved_drug_target_status.tsv',sep='\\t',index_col=0)\n",
    "\n",
    "with pd.ExcelWriter('../data/Supplementary_Data.xlsx') as xls:\n",
    "    gpcrs_with_families.to_excel(xls, sheet_name='Supplementary Table 1',index=False)\n",
    "    gpcrs_with_constraint.to_excel(xls, sheet_name='Supplementary Table 2',index=False)\n",
    "    gpcrs_with_phenotypes.to_excel(xls, sheet_name='Supplementary Table 3',index=False)\n",
    "    gpcrs_with_drug_targets.to_excel(xls, sheet_name='Supplementary Table 4',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.2 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
